src/zlib.zig

```diff
@@ -901,3 +945,277 @@ pub const ZlibCompressorArrayList = struct {
+                    this.mode = .GUNZIP;
+                    {
+                        const header = &[_]u8{ GZIP_HEADER_ID1, GZIP_HEADER_ID2 };
+                        state.next_in = header.ptr;
+                        state.avail_in = @intCast(header.len);
+                        var out: [0]u8 = undefined;
+                        state.avail_out = 0;
+                        state.next_out = &out; // passing a null pointer here causes it to return Z_STREAM_ERROR so we send zero-length instead.
+                        const ret = inflate(state, this.flush);
+                        bun.assert(ret == .Ok);
+                    }
+                } else {
+                    // There is no actual difference between INFLATE and INFLATERAW (after initialization).
+                    // the stream only partially matched the gzip header, bail.
+                    this.mode = .INFLATE;
+                }
+                return redd;
+            }
+
+            bun.assert(false); // invalid number of gzip magic number bytes read
+        }
+
+        // we're passed the header or there was no header. it is now safe to send everying to inflate().
+        this.do_inflate_loop = true;
+        if (!process_all_input) return bytes.len;
+        while (true) {
+            if (try this.doWork(output, this.flush)) {
+                break;
+            }
+        }
+        // bun.assert(state.avail_in == 0);
+
+        return bytes.len;
+    }
+
+    pub fn doWork(this: *ZlibDecompressorStreaming, output: *std.ArrayListUnmanaged(u8), flush: FlushValue) !bool {
+        const state = &this.state;
+        var out: [CHUNK]u8 = undefined;
+        const len = @min(CHUNK, this.chunkSize);
+        state.avail_out = len;
+        state.next_out = &out;
+
+        this.err = inflate(state, flush);
+        const ret = this.err;
+        bun.assert(ret != .StreamError);
+        if (ret == .NeedDict) return this.error_for_message((if (this.dictionary.len == 0) "Missing dictionary" else "Bad dictionary").ptr);
+        if (ret == .DataError) return this.error_for_message("Zlib error");
+        if (ret == .MemError) return error.ZlibError4;
+        const have = len - state.avail_out;
+        try output.appendSlice(bun.default_allocator, out[0..have]);
+        if (ret == .StreamEnd and this.mode == .GUNZIP and state.avail_in > 0 and state.next_in.?[0] != 0) {
+            _ = inflateReset(state);
+            return false;
+        }
+        if (ret == .StreamEnd) return true;
+        if (state.avail_out == 0) return false;
+        if (ret == .BufError and flush == .Finish) return this.error_for_message("unexpected end of file");
+        return true;
+    }
+
+    pub fn end(this: *ZlibDecompressorStreaming, output: *std.ArrayListUnmanaged(u8)) !void {
+        const state = &this.state;
+        state.next_in = null;
+        state.avail_in = 0;
+
+        const done = try this.doWork(output, this.finishFlush);
+        bun.assert(done);
```

Comment:
should this be panicking at runtime?

Reply:
it would be a bug on our part if it wasnt returning true in this location